src/NUnitFramework/framework/Internal/Execution/CompositeWorkItem.cs

```diff
@@ -243,7 +243,8 @@ private TestCommand MakeOneTimeTearDownCommand(List<SetUpTearDownItem> setUpTear
                 command = new OneTimeTearDownCommand(command, item);
 
             // Dispose of fixture if necessary
-            if (Test is IDisposableFixture && typeof(IDisposable).IsAssignableFrom(Test.TypeInfo.Type))
+            if (Test is IDisposableFixture && typeof(IDisposable).IsAssignableFrom(Test.TypeInfo.Type) &&
+                (Test as TestFixture)?.LifeCycle != LifeCycle.InstancePerTestCase)
                 command = new DisposeFixtureCommand(command);
```

Comment:
Here's a problem I didn't see at first. Let's say the fixture is `LifeCycle.InstancePerTestCase`. In that case, we won't construct it here. Now, let's say that the `OneTimeSetUp` is an instance method that sets some property in the fixture. It's going to throw an NRE.

So, it's really not a matter of choice whether we run `OneTimeSetUp` and `OneTimeTearDown` for each instance. It has to be done that way... unless there is something I'm missing.

I suggest adding a test case that does exactly that. Let it set some property of the fixture to a value. You should see the crash. Then you should fix it. :smiling_imp:

Reply:
i would really like to expand the test coverage on this
right now i'm basically hard-coding the LifeCycle flag to InstancePerTestCase then running the entire suite to check for bugs in this mode
since this affects everything, it needs to be tested, well - everywhere
i need to find an approach of testing this without duplicating every test code that was written
any ideas? like a global way to run all the tests with single then instance-per-test-case
what i'm going to do right now is simply find all the tests that fail and shouldn't with the InstancePerTestCase life cycle flag and add them to LifeCycleAttributeTests (basically duplicating the code.. )

but i know that's not ideal

Reply:
also about one time, i have a less obtrusive approach to fixing it - i can let MakeOneTimeSetUpCommand still construct the first instance, basically leaving it unchanged
FixturePerTestCaseCommand can then only construct the instance if its null

what do you think? currently it fixes all the failing test cases

Reply:
so i just had some free time and i pushed the latest change-set , i added more test cases went through the entire suite (with instance-per-test-case forced on) and made sure it makes sense on all test cases.

the only thing i would like to suggest is.. maybe we should throw an exception when `InstancePerTestCase` is used with an instance `OneTimeSetup/TearDown` method - this can lead to unexpected results since the instance is no longer shared

what do you think ?

Reply:
@CharliePoole 

> So, it's really not a matter of choice whether we run OneTimeSetUp and OneTimeTearDown for each instance. It has to be done that way... unless there is something I'm missing.

I do not believe that we should run OneTime methods more than one time. There's an inherent contradiction and we should error out the entire fixture and let the user decide how to resolve it, e.g. by moving the code into SetUp and TearDown. This is what I proposed in https://github.com/nunit/nunit/issues/2574.

Reply:
@jnm2 
Of course, that's up to you guys. I have no dog in this fight and was just trying to give you some information that might help.

It does seem to me that there are two contradictory things here and both of them can't be satisfied.

1. OneTimeSetUp definitely sounds like it should only run one time. :smile: 

2. If OneTimeSetUp is an instance method that changes state, it has to be run multiple times in order for the feature to be backward compatible.

Joseph, I'm used to the argument going the other way, with you being in favor of very strong backward compatibility. Why not here?

Reply:
> Joseph, I'm used to the argument going the other way, with you being in favor of very strong backward compatibility. Why not here?

The answer is that they opted in to a brand new feature by starting to use this new attribute. No one's code can break as a result of merely updating the referenced version of NUnit.

> UPDATE: OK, your inline comment clarified that for me. So if people use this feature, they have to have static onetime setup and teardown.

Or they have to change OneTimeSetUp to just SetUp, if that works for them. It might not be the same resolution for all situations.

Reply:
i guess we cant have our cake and eat it too, if you use this flag, something has to change in your codebase if you use OneTimeSetUp, i cant think of a reason of even using OneTimeSetUp when its per-instance anyway

either way, whatever you decide im here to do it ;)

Reply:
I would like this for C#8 nullability #3613 where the Constructor becomes SetUp and Dispose becomes the TearDown method. This allows then to use non-nullable fields and either far less forgive (!) operators in the code or the removal of #pragma. It also makes Roslyn Analyzer Disposal checks relevant.
method
OneTimeSetUp is sometimes used to create immutable data used by all tests or instantiate auxiliary classes needed. If that functionality can be maintained using a static method and fields would be great. A matching static OneTimeTearDown for cleaning would be also be needed.
If you need the https://github.com/nunit/nunit.analyzers to check for this, I'm willing to work on that there.

Reply:
@manfred-brands you could just use a static constructor for the one time setup in the scenario you described. I'm uncomfortable with the `OneTimeSetup` and `OneTimeTearDown` being allowed here because people may expect them to behave like the instance versions. Do you have a scenario for a static `OneTimeTearDown`?

Reply:
@rprouse We use `OneTimeSetUp` for expensive setups, like creating files(databases) which are then used by multiple tests. The `OneTimeTearDown` would then delete these. Although the tests would work if those methods are called for every test, they would certainly increase the time for the test run. So the behaviour would be the same. But it would not be  OneTime. The runner could verify that it is a static method which in turn means it can only use static fields etc.